Avoid unnecessary calls to the cloud provider

Miciah opened the pull request:
If the service controller is processing a service that does not have type LoadBalancer and does not have any external load balancer recorded in its status, do not ask the cloud provider to check for an associated load balancer.

**What this PR does / why we need it**:

We are seeing excessive `DescribeLoadBalancers` calls on some clusters.  This PR is intended to mitigate excessive API calls by avoiding unnecessary calls to the `DescribeLoadBalancers` API.

For example, these are the numbers from one of our production clusters, showing that calls to `DescribeLoadBalancers` dominate API usage:

| user           | API                   | calls |
| ---------------|---------------------- | ----- |
| cloud_provider | DescribeLoadBalancers | 23675 |
| ops_volume     | CreateTags            | 22653 |
| "              | DeleteSnapshot        | 21060 |
| "              | CreateSnapshot        | 12426 |
| "              | DescribeVolumes       | 11832 |
| cloud_provider | DescribeVolumes       |  1940 |
| "              | DescribeInstances     |  1546 |
| "              | AttachVolume          |   553 |
| "              | DetachVolume          |   514 |
| ops_monitoring | DescribeLoadBalancers |   373 |
| cloud_provider | DeleteVolume          |   310 |
| inventory      | DescribeInstances     |   270 |
| cloud_provider | DescribeSubnets       |   193 |
| "              | CreateVolume          |   192 |
| "              | CreateTags            |   185 |
| ops_volume     | DescribeInstances     |   156 |
| "              | DescribeSnapshots     |    83 |
| ops_config     | CreateTags            |    81 |
| "              | DescribeTags          |    81 |
| ops_monitoring | DescribeInstances     |     6 |
| s3_docker      | ListBuckets           |     2 |


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Related to #30700.

**Special notes for your reviewer**:

Based on #34271.

**Release note**:

```release-note
The service controller no longer checks for external load balancers associated with services that do not have type LoadBalancer and do not have any external load balancer recorded in their statuses.
```

Miciah committed:
Avoid unnecessary calls to the cloud provider

If the service controller is processing a service that does not have type
LoadBalancer and does not have any external load balancer recorded in its
status, do not call the cloud provider to check for an associated load
balancer.

dims commented:
/ok-to-test

dims commented:
/kind bug
/priority important-soon

dims commented:
/assign @andrewsykim

andrewsykim commented on pkg/controller/service/service_controller.go:
```diff
@@ -282,6 +282,11 @@ func (s *ServiceController) createLoadBalancerIfNeeded(key string, service *v1.S
 	var err error
 
 	if !wantsLoadBalancer(service) {
+		if v1helper.LoadBalancerStatusEqual(previousState, &v1.LoadBalancerStatus{}) {
```
Can we do `v1helper.LoadBalancerStatusEqual(previousState, &v1.LoadBalancerStatus{})` -> `v1helper.LoadBalancerStatusEqual(previousState, newState)` instead?

Miciah replied:
Using a literal here seems more explicit and less prone to future changes in how `newState` is initialized or used, but maybe I'm overthinking it.

If I do change this to `v1helper.LoadBalancerStatusEqual(previousState, newState)`, should I also move the assignment `newState = &v1.LoadBalancerStatus{}` from the end of the outer `if` block to precede that `if`block? Strictly speaking, we don't even need that assignment, so I assume it's there to make the intention more explicit.

Miciah replied:
Actually, that assignment is needed, and I would need to move it up in order to use it here.

andrewsykim replied:
Good point :)

andrewsykim commented:
@Miciah I have some concerns around assuming `LoadBalancerStatus` is the source of truth for whether a service has a loadbalancer or not. Right now we don't force any provider to return an updated loadbalancer status (see [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/service_controller.go#L312-L327)). It seems this change would mean: if a provider does not update `LoadBalancerStatus` then LB deletions (by changing the service type) would always be skipped, is that correct?

andrewsykim commented:
Having said that I think we should make updating `LoadBalancerStatus` a requirement (return err if not changed) but given it's not it might lead to some unexpected behaviour.

jhorwit2 commented:
@Miciah do you have numbers on how this PR helps reduce calls? Services should only go into this block of code and call the cloud when they pass the [`needsUpdate`](https://github.com/Miciah/kubernetes/blob/1954262d9d3d9665787082179ef8fc958f7b512a/pkg/controller/service/service_controller.go#L446) check which ignores services that aren't load balancers. From what I can tell, this will only reduce calls on service add events in the informer. Deletes and node updates go through different code paths.

Miciah commented:
@andrewsykim, 

> I have some concerns around assuming `LoadBalancerStatus` is the source of truth for whether a service has a loadbalancer or not. Right now we don't force any provider to return an updated loadbalancer status ([see here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/service_controller.go#L312-L327)). It seems this change would mean: if a provider does not update `LoadBalancerStatus` then LB deletions (by changing the service type) would always be skipped, is that correct?

I believe that is correct, although we do log the error so an operator can handle the situation appropriately.  I would even argue that logging (and leaving resolution to the operator) may be better than unconditionally deleting a load balancer that may have been created outside the controller.

> Having said that I think we should make updating `LoadBalancerStatus` a requirement (return err if not changed), though that will require more work. Given it's not, it might lead to some unexpected behaviour.

I would agree, except that we used to return `err` if the `persistUpdate` failed, and https://github.com/kubernetes/kubernetes/pull/56370 explicitly changed that behavior.  Quoting from that PR:

> There was only one case where we didn't want to retry, which was when we updated the service status since that handled retries itself. In that case I moved to the common pattern of using `runtime.HandleError`.

`persistUpdate` does retry but only a limited number of times.  Retrying and eventually giving up in `persistUpdate` means we can lose status updates.  Maybe the reasoning is that we don't want a bunch of status-update failures to fill the work queue.  Otherwise, if that is not a concern, could we delete the retry logic from `persistUpdate` and return the `err` so that it gets retried at a higher level?

Digging further back in history, `cachedService` used to have an `appliedState` field.  It was deleted in https://github.com/kubernetes/kubernetes/pull/25189, but it is not obvious to me whether there was a specific reason to avoid storing `appliedState` in `cachedService`.  If not, is there any reason I should not re-introduce `appliedState` and using it for the comparison in `createLoadBalancerIfNeeded`?

Thanks for your help on this PR!

Miciah commented:
@jhorwit2,

> do you have numbers on how this PR helps reduce calls?

Not yet.  I am working downstream to get the fix onto one or more of our large clusters so that we can get those numbers to justify merging upstream.

> Services should only go into this block of code and call the cloud when they pass the [needsUpdate](https://github.com/Miciah/kubernetes/blob/1954262d9d3d9665787082179ef8fc958f7b512a/pkg/controller/service/service_controller.go#L446) check which ignores services that aren't load balancers. From what I can tell, this will only reduce calls on service add events in the informer.

I believe you are correct.  Our use-case may be unusually highly multi-tenant and developer-oriented and  thus have more service-add events than the typical cluster.

deads2k commented:
@Miciah nag. This is an outstanding pick in the list.

soltysh commented:
/remove-lifecycle rotten
@Miciah gentle nag :smiley_cat:

andrewsykim commented:
/remove-lifecycle rotten

> Not yet. I am working downstream to get the fix onto one or more of our large clusters so that we can get those numbers to justify merging upstream.

@Miciah any updates on this? Were you able to get some numbers to justify if an optimization here is needed? If so & the fix is trivial let's try to get something in for v1.14 (code freeze is Mar 7th), otherwise let's revisit this for v1.15.

Miciah commented:
Here are before and after numbers showing how many calls to various AWS APIs one of our production clusters made over the course of a day.  The numbers for various APIs may be affected by other fixes or configuration changes that were rolled out at the same time, but we did see a substantial drop in the number of `DescribeLoadBalancers` calls.

Before:

| API                          | calls |
|---------------------------- |----- |
| DescribeLoadBalancers        | 26484 |
| CreateTags                   | 14335 |
| DescribeVolumes              | 10454 |
| DeleteSnapshot               | 9724  |
| CreateSnapshot               | 7345  |
| DescribeInstances            | 1758  |
| DescribeTags                 | 1338  |
| DescribeSubnets              | 580   |
| DetachVolume                 | 398   |
| AttachVolume                 | 370   |
| DescribeVolumeStatus         | 161   |
| DescribeVolumesModifications | 36    |
| DescribeSnapshots            | 32    |
| DescribeAlarms               | 28    |
| DescribeAddresses            | 25    |
| DescribeAvailabilityZones    | 24    |
| DescribeInstanceStatus       | 17    |
| DescribeAccountAttributes    | 8     |
| DescribeSecurityGroups       | 6     |
| DescribeLaunchTemplates      | 6     |
| DeleteNetworkInterface       | 6     |
| DescribePlacementGroups      | 4     |
| DescribeKeyPairs             | 4     |
| DescribeHosts                | 4     |
| ListAliases                  | 3     |
| ListBuckets                  | 2     |
| DescribeIdFormat             | 2     |
| DescribeVolumeAttribute      | 1     |
| CreateNetworkInterface       | 1     |

After:

| API                          | calls |
|---------------------------- |----- |
| DescribeInstances            | 17727 |
| DescribeVolumes              | 590   |
| DescribeSubnets              | 575   |
| DescribeLoadBalancers        | 391   |
| DescribeVolumeStatus         | 11    |
| DescribeAddresses            | 9     |
| DescribeAvailabilityZones    | 6     |
| DescribeSnapshots            | 5     |
| DescribeLaunchTemplates      | 5     |
| DescribeAlarms               | 4     |
| DescribeAccountAttributes    | 4     |
| DeleteNetworkInterface       | 4     |
| DescribeTags                 | 3     |
| DescribeInstanceStatus       | 3     |
| DescribeSecurityGroups       | 2     |
| DescribePlacementGroups      | 2     |
| DescribeKeyPairs             | 2     |
| DescribeHosts                | 2     |
| DescribeLaunchConfigurations | 1     |
| DescribeIdFormat             | 1     |
| DescribeAutoScalingGroups    | 1     |
| CreateNetworkInterface       | 1     |

andrewsykim commented:
Thanks for sharing those numbers with us! More information on how the cluster was configured & activity on those clusters would be helpful to make sure we are measuring the right thing. 

One thing I'm confused about is why the changes in this PR reduces API calls for `DescribeLoadBalancers` given we have a pre-check for if a Service needs to be updated prior to adding it to the service queue (see https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/service_controller.go#L445-L446). 

Based on the diff in this PR, if there is a service that doesn't want a load balancer anymore, we also make sure it didn't already have an LB:
```go
if !wantsLoadBalancer(service) {
		if v1helper.LoadBalancerStatusEqual(previousState, &v1.LoadBalancerStatus{}) {
			return nil
		}
                ...
```

But this would only trigger a lot of API calls if you have a lot of services going from Type=LoadBalancer to another type. Otherwise, it would be caught in the pre-check as part of [needsUpdate](https://github.com/kubernetes/kubernetes/blame/master/pkg/controller/service/service_controller.go#L444-L500).

thockin commented:
I know this PR has been languishing, but I feel the need to point out that this is part of the reconciliation process that makes Kubernetes resilient.  

If we drop this, we introduce the possibility of leaking load-balancers in the cloud provider.  E.g. consider what happens if the controller-manager crashes or is restarted in the middle of one of these loops.  The status gets cleared, the type is now ClusterIP, but a cloud LB exists.  If we do not try to reconcile, that LB will exist forever.

Miciah commented:
To close the loop:  I believe this change is no longer needed, thanks to https://github.com/kubernetes/kubernetes/pull/78262/commits/aa3f81d657c1d4b7c789352af7179c08508adc3c.

To restate the original problem:  At the time I wrote this PR, the service controller added every newly created service to its work queue, which was also used for update and delete events.  When a worker for this queue subsequently processed the service, it would check if the service needed an external load-balancer, and if not, it would then make a `GetLoadBalancer` call to check if a load balancer existed for the service, in case the service had been added to the queue by an update that changed its type from "LoadBalancer", in which case any previously provisioned load-balancer would need to be deleted.

This PR adds a check to skip the `GetLoadBalancer` call if the service's status indicates that it has no
associated load-balancer, which should always be the case for a newly created service if its type is not "LoadBalancer".

However, #78262 modifies the service controller's logic to add a newly created service to the work queue only if its type is "LoadBalancer", which obviates the need for the check that this PR adds.
